File: clients/python/tests/chunking/resources/csharp/Sample.cs
/**
 * Sample C# file for testing code chunking.
 *
 * This file contains various C# constructs to test AST parsing.
 */

using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Sample
{
    /// <summary>
    /// Constants for the application.
    /// </summary>
    public static class Constants
    {
        public const int MaxRetries = 3;
        public const double DefaultTimeout = 30.0;
    }

    /// <summary>
    /// Represents a user in the system.
    /// </summary>
    public class User
    {
        public string Id { get; }
        public string Name { get; }
        public string? Email { get; set; }

        public User(string id, string name)
        {
            Id = id;
            Name = name;
        }

        public User(string id, string name, string email) : this(id, name)
        {
            Email = email;
        }

        public string GetDisplayName()
        {
            return Name ?? Email ?? Id;
        }
    }

    /// <summary>
    /// Interface for services.
    /// </summary>
    public interface IService
    {
        Task InitializeAsync();
        bool IsReady { get; }
    }

    /// <summary>
    /// Custom exception for service errors.
    /// </summary>
    public class ServiceException : Exception
    {
        public ServiceException(string message) : base(message) { }
        public ServiceException(string message, Exception inner) : base(message, inner) { }
    }

    /// <summary>
    /// Service for managing users.
    /// </summary>
    public class UserService : IService
    {
        private readonly Dictionary<string, User> _users = new();
        private bool _initialized;

        public bool IsReady => _initialized;

        public async Task InitializeAsync()
        {
            await Task.Delay(10); // Simulate async initialization
            _initialized = true;
        }

        public User CreateUser(string id, string name, string? email = null)
        {
            if (string.IsNullOrEmpty(id))
            {
                throw new ServiceException("ID cannot be empty");
            }

            var user = email != null
                ? new User(id, name, email)
                : new User(id, name);

            _users[id] = user;
            OnUserCreated(user);
            return user;
        }

        public User? GetUser(string id)
        {
            return _users.TryGetValue(id, out var user) ? user : null;
        }

        public bool DeleteUser(string id)
        {
            return _users.Remove(id);
        }

        private void OnUserCreated(User user)
        {
            // Internal callback
        }
    }

    /// <summary>
    /// Generic container class.
    /// </summary>
    public class Container<T>
    {
        public T Value { get; set; }

        public Container(T value)
        {
            Value = value;
        }
    }

    /// <summary>
    /// Record type for immutable data.
    /// </summary>
    public record UserRecord(string Id, string Name, string? Email = null);

    /// <summary>
    /// Struct for value types.
    /// </summary>
    public struct Point
    {
        public double X { get; init; }
        public double Y { get; init; }

        public double Distance => Math.Sqrt(X * X + Y * Y);
    }

    /// <summary>
    /// Enum for service status.
    /// </summary>
    public enum ServiceStatus
    {
        Pending,
        Running,
        Stopped,
        Error
    }

    /// <summary>
    /// Utility class with static methods.
    /// </summary>
    public static class MathUtils
    {
        /// <summary>
        /// Calculate factorial of n.
        /// </summary>
        public static long CalculateFactorial(int n)
        {
            if (n <= 1) return 1;
            return n * CalculateFactorial(n - 1);
        }

        /// <summary>
        /// Fetch data asynchronously.
        /// </summary>
        public static async Task<Dictionary<string, string>> FetchDataAsync(string url)
        {
            await Task.Delay(10); // Simulate network call
            return new Dictionary<string, string>
            {
                ["url"] = url,
                ["status"] = "ok"
            };
        }

        /// <summary>
        /// Process items with optional validation.
        /// </summary>
        public static List<string> ProcessItems(List<string> items, bool validate = true)
        {
            var result = new List<string>();
            foreach (var item in items)
            {
                if (validate && string.IsNullOrEmpty(item))
                    continue;
                result.Add(item.Trim().ToLower());
            }
            return result;
        }
    }

    /// <summary>
    /// Extension methods for strings.
    /// </summary>
    public static class StringExtensions
    {
        public static string Truncate(this string value, int maxLength)
        {
            if (string.IsNullOrEmpty(value)) return value;
            return value.Length <= maxLength ? value : value[..maxLength];
        }
    }

    /// <summary>
    /// Main program entry point.
    /// </summary>
    public class Program
    {
        public static async Task Main(string[] args)
        {
            var service = new UserService();
            await service.InitializeAsync();

            var user = service.CreateUser("1", "Test User", "test@example.com");
            Console.WriteLine($"Created user: {user.GetDisplayName()}");

            var result = MathUtils.CalculateFactorial(5);
            Console.WriteLine($"Factorial: {result}");
        }
    }
}

File: clients/python/tests/chunking/resources/haskell/Sample.hs
{- |
Module      : Sample
Description : Sample Haskell file for testing code chunking
Copyright   : (c) ProximaDB Team, 2025
License     : Apache-2.0

This module contains various Haskell constructs to test AST parsing.
-}

module Sample
    ( -- * Constants
      maxRetries
    , defaultTimeout
      -- * Data Types
    , User(..)
    , ServiceError(..)
    , ServiceStatus(..)
      -- * Type Classes
    , Service(..)
    , Displayable(..)
      -- * User Operations
    , createUser
    , getDisplayName
    , setEmail
      -- * UserService
    , UserService
    , newUserService
    , addUser
    , getUser
    , deleteUser
      -- * Utility Functions
    , calculateFactorial
    , fibonacci
    , processItems
    , withRetry
    ) where

import qualified Data.Map.Strict as Map
import Data.Map.Strict (Map)
import Control.Monad (when)
import Control.Exception (Exception, throw, catch, SomeException)
import Data.Maybe (fromMaybe)

-- | Maximum number of retries for operations
maxRetries :: Int
maxRetries = 3

-- | Default timeout in seconds
defaultTimeout :: Double
defaultTimeout = 30.0

-- | Represents a user in the system
data User = User
    { userId    :: !String
    , userName  :: !String
    , userEmail :: !(Maybe String)
    } deriving (Show, Eq)

-- | Custom exception for service errors
data ServiceError
    = NotFoundError String
    | InvalidInputError String
    | InternalError String
    deriving (Show, Eq)

instance Exception ServiceError

-- | Service status enumeration
data ServiceStatus
    = Pending
    | Running
    | Stopped
    | Error String
    deriving (Show, Eq)

-- | Type class for displayable objects
class Displayable a where
    -- | Get display representation
    display :: a -> String

instance Displayable User where
    display = getDisplayName

-- | Type class for services
class Service s where
    -- | Initialize the service
    initialize :: s -> IO s

    -- | Check if service is ready
    isReady :: s -> Bool

-- | Create a new user
createUser :: String -> String -> Maybe String -> Either ServiceError User
createUser uid name email
    | null uid  = Left $ InvalidInputError "ID cannot be empty"
    | otherwise = Right $ User uid name email

-- | Get display name for a user
getDisplayName :: User -> String
getDisplayName user =
    case userName user of
        "" -> fromMaybe (userId user) (userEmail user)
        n  -> n

-- | Set user email
setEmail :: String -> User -> User
setEmail email user = user { userEmail = Just email }

-- | User service data type
data UserService = UserService
    { users       :: !(Map String User)
    , initialized :: !Bool
    } deriving (Show)

-- | Create a new UserService
newUserService :: UserService
newUserService = UserService Map.empty False

instance Service UserService where
    initialize svc = return $ svc { initialized = True }
    isReady = initialized

-- | Add a user to the service
addUser :: String -> String -> Maybe String -> UserService -> Either ServiceError UserService
addUser uid name email svc
    | null uid = Left $ InvalidInputError "ID cannot be empty"
    | otherwise =
        let user = User uid name email
            newUsers = Map.insert uid user (users svc)
        in Right $ svc { users = newUsers }

-- | Get a user by ID
getUser :: String -> UserService -> Maybe User
getUser uid svc = Map.lookup uid (users svc)

-- | Delete a user by ID
deleteUser :: String -> UserService -> (Bool, UserService)
deleteUser uid svc =
    let exists = Map.member uid (users svc)
        newUsers = Map.delete uid (users svc)
    in (exists, svc { users = newUsers })

-- | Calculate factorial of n
calculateFactorial :: Integer -> Integer
calculateFactorial n
    | n <= 1    = 1
    | otherwise = n * calculateFactorial (n - 1)

-- | Calculate nth Fibonacci number
fibonacci :: Int -> Integer
fibonacci n = fibs !! n
  where
    fibs = 0 : 1 : zipWith (+) fibs (tail fibs)

-- | Process items with optional validation
processItems :: Bool -> [String] -> [String]
processItems validate items =
    let filtered = if validate
                   then filter (not . null) items
                   else items
    in map (map toLowercase . trim) filtered
  where
    toLowercase c
        | c >= 'A' && c <= 'Z' = toEnum (fromEnum c + 32)
        | otherwise = c
    trim = dropWhile (== ' ') . reverse . dropWhile (== ' ') . reverse

-- | Execute an action with retry
withRetry :: Int -> IO a -> IO (Either SomeException a)
withRetry maxTries action = go maxTries
  where
    go 0 = return $ Left $ toException $ InternalError "Max retries exceeded"
    go n = catch (Right <$> action) handler
      where
        handler :: SomeException -> IO (Either SomeException a)
        handler e = go (n - 1)

    toException :: Exception e => e -> SomeException
    toException = toException

-- | Higher-order function example
mapUsers :: (User -> a) -> UserService -> [a]
mapUsers f svc = map f $ Map.elems (users svc)

-- | Generic container
data Container a = Container
    { containerValue :: a
    } deriving (Show, Eq, Functor)

-- | Type alias
type UserId = String
type UserMap = Map UserId User

-- | Newtype wrapper
newtype Email = Email { unEmail :: String }
    deriving (Show, Eq)

-- | Pattern synonym
pattern AdminUser :: String -> User
pattern AdminUser name = User "admin" name Nothing

-- | Main entry point
main :: IO ()
main = do
    let svc = newUserService
    svc' <- initialize svc

    case addUser "1" "Test User" (Just "test@example.com") svc' of
        Left err -> putStrLn $ "Error: " ++ show err
        Right svc'' -> do
            case getUser "1" svc'' of
                Nothing -> putStrLn "User not found"
                Just user -> putStrLn $ "Created user: " ++ getDisplayName user

    let result = calculateFactorial 5
    putStrLn $ "Factorial: " ++ show result
